arangod/Futures/SharedState.h

```diff
@@ -0,0 +1,292 @@
+  SharedState& operator=(SharedState&&) = delete;
+  
+  /// True if state is OnlyCallback or Done.
+  /// May call from any thread
+  bool hasCallback() const noexcept {
+    auto const state = _state.load(std::memory_order_acquire);
+    return state == State::OnlyCallback || state == State::Done;
+  }
+  
+  /// True if state is OnlyResult or Done.
+  /// May call from any thread
+  bool hasResult() const noexcept {
+    auto const state = _state.load(std::memory_order_acquire);
+    return state == State::OnlyResult || state == State::Done;
+  }
+  
+  /// Call only from consumer thread (since the consumer thread can modify the
+  ///   referenced Try object; see non-const overloads of `future.result()`,
+  ///   etc., and certain Future-provided callbacks which move-out the result).
+  ///
+  /// Unconditionally returns a reference to the result.
+  ///
+  /// State dependent preconditions:
+  ///
+  /// - Start or OnlyCallback: Never safe - do not call. (Access in those states
+  ///   would be undefined behavior since the producer thread can, in those
+  ///   states, asynchronously set the referenced Try object.)
+  /// - OnlyResult: Always safe. (Though the consumer thread should not use the
+  ///   returned reference after it attaches a callback unless it knows that
+  ///   the callback does not move-out the referenced result.)
+  /// - Done: Safe but sometimes unusable. (Always returns a valid reference,
+  ///   but the referenced result may or may not have been modified, including
+  ///   possibly moved-out, depending on what the callback did; some but not
+  ///   all callbacks modify (possibly move-out) the result.)
+  Try<T>& getTry() {
+    TRI_ASSERT(hasResult());
+    return _result;
+  }
+  Try<T> const& getTry() const {
+    TRI_ASSERT(hasResult());
+    return _result;
+  }
+  
+  /// Call only from consumer thread.
+  /// Call only once - else undefined behavior.
+  ///
+  /// See FSM graph for allowed transitions.
+  ///
+  /// If it transitions to Done, synchronously initiates a call to the callback,
+  /// and might also synchronously execute that callback (e.g., if there is no
+  /// executor or if the executor is inline).
+  template <typename F>
+  void setCallback(F&& func) {
+    TRI_ASSERT(!hasCallback());
+    
+    // construct _callback first; if that fails, context_ will not leak
+    _callback = std::forward<F>(func);
+    //::new (&_callback) Callback(std::forward<F>(func));
+    
+    auto state = _state.load(std::memory_order_acquire);
+    while (true) {
+      switch (state) {
+        case State::Start:
+          if (_state.compare_exchange_strong(state, State::OnlyCallback, std::memory_order_release)) {
+            return;
+          }
+          TRI_ASSERT(state == State::OnlyResult); // race with setResult
+          [[gnu::fallthrough]];
+          
+        case State::OnlyResult:
+          if (_state.compare_exchange_strong(state, State::Done, std::memory_order_release)) {
```

Comment:
This CAS has to use memory_order_acquire to synchronize-with the release-CAS in line 202.